test/SemVer.unittest.js

```diff
@@ -0,0 +1,513 @@
+	});
+
+	describe("parseRange", () => {
+		const cases = {
+			"=3": ["3", "v3", "3.x", "3.X", "3.x.x", "3.*", "3.*.*", "^3", "^3.x"],
+			"=3.0": ["3.0", "v3.0", "3.0.x", "3.0.X", "3.0.*", "~3.0"],
+			"^3.4": ["^3.4.*"],
+			"3.4 - 6.5": [">=3.4 <=6.5"],
+			"<=3.4": ["<3.4 || =3.4"],
+			">3.4": [">=3.4 !3.4"]
+		};
+		for (const key of Object.keys(cases)) {
+			describe(key, () => {
+				for (const c of cases[key])
+					it(`should be equal ${c}`, () => {
+						expect(parseRange(c)).toEqual(parseRange(key));
+					});
+			});
+		}
+	});
+
+	describe("rangeToString", () => {
+		const cases = {
+			"1": "^1",
+			"1.2": "~1.2",
+			"1.2.3": "=1.2.3",
+			"^1.2.3": "^1.2.3",
+			"~1.2.3": "~1.2.3",
+			"0.0.1": "=0.0.1",
+			"^0.0.1": "=0.0.1",
+			"^0.1.2": "~0.1.2",
+			"~0.0.1": "~0.0.1",
+			"~0.1.2": "~0.1.2",
+			">=1.2.3": ">=1.2.3",
+			"1.2.3-beta.25": "=1.2.3-beta.25",
+			"1.2.3-beta.25+12.34": "=1.2.3-beta.25+12.34",
+			"1.2.3+12.34": "=1.2.3+12.34",
+			">=1.2.3-beta.25": ">=1.2.3-beta.25",
+			">=1.2.3-beta.25+12.34": ">=1.2.3-beta.25+12.34",
+			">=1.2.3+12.34": ">=1.2.3+12.34",
+			"<1.2.3-beta.25": "<1.2.3-beta.25",
+			"<1.2.3-beta.25+12.34": "<1.2.3-beta.25+12.34",
+			"<1.2.3+12.34": "<1.2.3+12.34",
+			"1.2.3 - 3.2.1": ">=1.2.3 (<3.2.1 || =3.2.1)",
+			">3.4": ">=3.4 not(~3.4)",
+			"1 || 2 || 3": "^1 || ^2 || ^3",
+			"1.2.3 - 3.2.1 || >3 <=4 || 1":
+				">=1.2.3 (<3.2.1 || =3.2.1) || >=3 not(^3) (<4 || ^4) || ^1"
+		};
+		for (const key of Object.keys(cases)) {
+			const expected = cases[key];
+			it(`should ${key} stringify to ${expected}`, () => {
+				expect(rangeToString(parseRange(key))).toEqual(expected);
+			});
+		}
+	});
+
+	describe("satisfies", () => {
+		const cases = {
+			// table cases
+			">=1": [
+				"1",
+				"2",
+				"!1-beta",
+				"!2-beta",
+				"1.2",
+				"!1.2-beta",
+				"2.2",
+				"!2.2-beta",
+				"1.beta",
+				"!1.beta-beta",
+				"!2.beta-beta"
+			],
+			">=1-beta": [
+				"1",
+				"2",
+				"1-beta",
+				"1-gamma",
+				"!1-alpha",
+				"!2-beta",
+				"1.2",
+				"!1.2-beta",
+				"2.2",
+				"!2.2-beta",
+				"1.beta",
+				"!1.beta-beta",
+				"2.beta",
+				"!2.beta-beta"
+			],
+			">=1.2": [
+				"!1",
+				"2",
+				"!1-beta",
+				"!2-beta",
+				"!1.1",
+				"1.2",
+				"1.3",
+				"2.1",
+				"2.2",
+				"2.3",
+				"1.beta",
+				"2.beta"
+			],
+			"~1.2": [
+				"!1",
+				"!2",
+				"!10",
+				"!1-beta",
+				"!2-beta",
+				"!1.1",
+				"1.2",
+				"!1.3",
+				"!1.20"
+			],
+			">=1.beta": [
+				"!1",
+				"2",
+				"!1-beta",
+				"!2-beta",
+				"!1.2",
+				"2.2",
+				"!1.0",
+				"!1.100",
+				"!1.alpha",
+				"1.beta",
+				"1.gamma",
+				"2.beta"
+			],
+			// fixed cases
+			"2": [
+				"2",
+				"2.0.0",
+				"2.99.99",
+				"!2.3.4-beta",
+				"!2.3.4-beta.1",
+				"!2.3.4-beta+123",
+				"2.3.4+123",
+				"!1",
+				"!1.2.3",
+				"!3",
+				"!3.4.5"
+			],
+			"1.2.3-beta.1.2+33.44": [
+				"1.2.3-beta.1.2+33.44",
+				"!1.2.3-beta.1.2+33.44.55",
```

Comment:
Semver (and the semver package in npm) is pretty clear that this should be contained in the range, and that build strings should *not* influence precedence. From https://semver.org/, section 10:

> Build metadata MUST be ignored when determining version precedence. Thus two versions that differ only in the build metadata, have the same precedence.

From nodejs:

```
> var s = require("semver")
> s.satisfies('1.2.3-beta.1.2+0', new s.Range('1.2.3-beta.1.2+33.44'))
true
> s.satisfies('1.2.3-beta.1.2+33.44.55', new s.Range('1.2.3-beta.1.2+33.44'))
true
```
If you're going for full compatibility, I'd suggest stripping the build from the precedence calculations.

Reply:
One easy way to do this is to consider the start of build data as the end of the array case when doing comparisons?

Reply:
ok. That would simplify the logic a little bit. I would omit build version during parsing the ranges.